src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -775,41 +772,60 @@ async_manifest_view::time_based_retention(
             bool eof = false;
 
             auto cursor = std::move(res.value());
-            while (!truncation_point && !eof) {
-                cursor->with_manifest([this, boundary, &truncation_point](
-                                        const partition_manifest& manifest) {
-                    for (const auto& meta : manifest) {
-                        if (
-                          meta.base_offset > _stm_manifest.get_start_offset()) {
-                            break;
-                        }
+            while (
+              cursor->get_status()
+              == async_manifest_view_cursor_status::materialized_spillover) {
+                auto eof = cursor->with_manifest(
+                  [this, boundary, &result](
+                    const partition_manifest& manifest) {
+                      if (
+                        _stm_manifest.get_start_offset()
+                        == manifest.get_start_offset()) {
+                          // We reached the STM manifest
+                          return true;
+                      }
```

Comment:
Given the condition in the while, `status == async_manifest_view_cursor_status::materialized_spillover`, I don't think branch can be reached. In any case, this code has to include the stm manifest in the iteration for the case when the first segment with a timestamp greater than the boundary is the first one in the stm manifest. The previous code accounted for this.

Reply:
fixed that